### PR TITLE
Bug 1853309 - Expose Creating Shopping Analysis API

### PR DESCRIPTION
**Pull request by calumozilla**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1853309

### CODE REVIEW

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -760,6 +760,39 @@ class GeckoEngineSession(
         })
     }
 
+    /**
+     * See [EngineSession.requestCreateProductAnalysis]
+     */
+    @Suppress("ComplexCondition")
```
Comment (rahulsainani) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:766]: This shouldn't be required

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 2]
```diff
@@ -871,6 +871,18 @@ abstract class EngineSession(
         onException: (Throwable) -> Unit,
     )
 
+    /**
+     * Requests the creation of product analysis for a given product page URL.
+     *
+     * @param onSuccess callback invoked if the engine API returns a valid response.
+     * @param onError callback invoked if there was an error getting the response.
+     */
+    abstract fun requestCreateProductAnalysis(
```
Comment (rahulsainani) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:880]: AFAIU this is the analyze API, should we name it `analyzeProduct` or `reAnalyzeProduct` here so it's more clear?

**Comment by rahulsainani** [comment]
Can you please rebase so this uses the latest GV changes and we can test this patch with integration on Fenix side

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 3]
```diff
@@ -760,6 +760,39 @@ class GeckoEngineSession(
...
+    override fun requestCreateProductAnalysis(
+        url: String,
+        onResult: (String) -> Unit,
+        onException: (Throwable) -> Unit,
+    ) {
+        geckoSession.requestCreateAnalysis(url).then({
+                response ->
+            if (response == null) {
```
Comment (rahulsainani) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:774]: ❓ When can response be null?

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (calumozilla) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Removed Suppress

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 2 above)
Comment (calumozilla) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: renamed to reanalyzeProduct

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 3 above)
Comment (calumozilla) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Removed null check, GV should return NonNull value

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 4]
```diff
@@ -863,15 +863,39 @@ abstract class EngineSession(
...
 
+    /**
+     * Requests the reanalysis of a product for a given product page URL.
+     *
+     * @param onResult callback invoked if the engine API returns a valid response.
+     * @param onException callback invoked if there was an error getting the response.
+     */
+    abstract fun reanalyzeProduct(
+        url: String,
+        onResult: (String?) -> Unit,
```
Comment (rahulsainani) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:883]: The callback should return a non nullable String as GV is returning non null when there's a valid response. Same for the function below.

**Review by rahulsainani — APPROVED** [review]
Overall it looks good and works well. 
Left one comment about non nullable string for both the functions. 
Do we have a bug that captures addition of tests for these?

**Comment by calumozilla** [comment]
> Overall it looks good and works well. Left one comment about non nullable string for both the functions. Do we have a bug that captures addition of tests for these?

Created a bug for automated tests, since this was done manually: https://bugzilla.mozilla.org/show_bug.cgi?id=1854141

**Review by rahulsainani — APPROVED** [review]

**Pull request merged**